### PR TITLE
UCT: Error handling API

### PR DESCRIPTION
**Pull request by alex--m**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1661/ for details.

**Comment by alex--m** [comment]
@yosefe Initial API changeset towards error handling... let's discuss.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1662/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1049/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1050/ for details (Mellanox internal link).

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -76,7 +77,7 @@ ucs_status_t ucp_ep_add_pending_uct(ucp_ep_h ep, uct_ep_h uct_ep,
     }
 
     /* Forced progress */
-    status = req->func(req);
+    status = req->func(req, status);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: why do you alway want to pass UCS_ERR_BUSY to req->func?

File: src/ucp/tag/eager_snd.c [hunk 2]
```diff
@@ -270,46 +274,48 @@ static ucs_status_t ucp_tag_eager_contig_zcopy_multi(uct_pending_req_t *self)
...
-    ucs_status_t status;
-
     status = ucp_do_am_bcopy_single(self, UCP_AM_ID_EAGER_ONLY,
                                     ucp_tag_pack_eager_only_generic);
     if (status != UCS_OK) {
         return status;
     }
 
-    ucp_tag_eager_generic_complere(self);
+    ucp_tag_eager_generic_complere(self, status);
```
Comment (abouteiller) [inline src/ucp/tag/eager_snd.c]: typo complere

File: src/uct/api/tl.h [hunk 3]
```diff
@@ -59,6 +51,14 @@ typedef struct uct_iface_ops {
     int          (*iface_is_reachable)(uct_iface_h iface,
                                        const uct_iface_addr_t *addr);
 
+    void         (*ep_destroy)(uct_ep_h ep);
```
Comment (brminich) [inline src/uct/api/tl.h]: I'd add some comment that function pointers in uct_iface_ops should be in strict order: all functions with iface as a first parameter must come first. Also some test with uct_base_iface_set_error function would be good to have, because it seem to be quite easy to introduce an error by adding some new function to uct_iface_ops in a wrong place.

File: src/uct/base/uct_iface.c [hunk 4]
```diff
@@ -150,6 +150,36 @@ static ucs_status_t uct_base_iface_flush(uct_iface_h tl_iface)
...
+{
+    return ep->iface->status;
+}
+
+void uct_base_iface_set_error(uct_iface_h tl_iface, ucs_status_t status)
+{
+    void **func_iter;
+    tl_iface->status = status;
+
+    /* Overwrite functions with endpoint as first argument */
```
Comment (brminich) [inline src/uct/base/uct_iface.c]: minor: seems to be a typo in the comment, because functions with iface argument are from the beginning.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1688/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1110/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1713/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1714/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1111/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1716/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1113/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1114/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1718/ for details.

**Comment by alex--m** [comment]
@shamisp - could you please review?

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -795,6 +795,8 @@ void ucp_ep_destroy(ucp_ep_h ep);
  * prior to this call are completed both at the origin and at the target
  * @ref ucp_ep_h "endpont" when this call returns.
  *
+ * @note Errors originating from non-blocking calls may return from invocation.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: The note does not read right. I would suggest something like:
`The call flushes outstanding communication operations and therefore it may return error codes associated with the operations. `

Does it sound right ?

**Comment by shamisp** [comment]
the code itself looks fine, but the documentation has to improved

File: src/ucp/core/ucp_ep.c (hunk 1 above)
Comment (abouteiller) [inline src/ucp/core/ucp_ep.c]: More generally, req->func is called with status = {UCS_OK or UCS_ERR_BUSY}. It looks like it should produce a status, but not accept one as input.

File: src/ucp/proto/proto_am.c [hunk 6]
```diff
@@ -23,12 +23,13 @@ static size_t ucp_proto_pack(void *dest, void *arg)
     return 0;
 }
 
-ucs_status_t ucp_proto_progress_am_bcopy_single(uct_pending_req_t *self)
+ucs_status_t ucp_proto_progress_am_bcopy_single(uct_pending_req_t *self,
+                                                ucs_status_t status)
```
Comment (abouteiller) [inline src/ucp/proto/proto_am.c]: Why does these operations take an input status?

File: src/ucs/type/status.h [hunk 7]
```diff
@@ -42,9 +42,23 @@ typedef enum {
     UCS_ERR_TIMED_OUT              = -20,
     UCS_ERR_EXCEEDS_LIMIT          = -21,
     UCS_ERR_UNSUPPORTED            = -22,
+
```
Comment (abouteiller) [inline src/ucs/type/status.h]: Some of the already existing errors could be considered "link errors" (TIMED_OUT, UNREACHABLE? ) 

ERR_SHMEM_SEGMENT: why does this error code specifically name an UCX client? Could it be "anonymized" ?

File: src/ucp/proto/proto_am.c (hunk 6 above)
Comment (alex--m) [inline src/ucp/proto/proto_am.c]: In UCT the uct_pending_callback_t prototype has been added the status argument for the case where during the time since posting the request an error has occured and progressing it is no longer possible.

File: src/ucs/type/status.h (hunk 7 above)
Comment (alex--m) [inline src/ucs/type/status.h]: You are right. The reason I didn't partition the existing errors is that some errors may end up in more than one group, and making these changes will be part of the implementation - once the API will reach consensus.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1144/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1751/ for details.

File: src/ucp/proto/proto_am.c (hunk 6 above)
Comment (alex--m) [inline src/ucp/proto/proto_am.c]: Further discussion with yossi led to the decision to remove the status argument from pending callbacks. I'll make the change and resubmit.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1756/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1148/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1759/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1151/ for details (Mellanox internal link).

File: src/ucp/tag/eager_snd.c [hunk 8]
```diff
@@ -293,7 +295,7 @@ static ucs_status_t ucp_tag_eager_generic_single(uct_pending_req_t *self)
         return status;
     }
 
-    ucp_tag_eager_generic_complere(self);
+    ucp_tag_eager_generic_complete(self, UCS_OK);
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: shouldn't status be passed here instead of UCS_OK?

File: src/ucp/tag/eager_snd.c [hunk 9]
```diff
@@ -270,13 +270,15 @@ static ucs_status_t ucp_tag_eager_contig_zcopy_multi(uct_pending_req_t *self)
                                  ucp_tag_eager_contig_zcopy_req_complete);
 }
 
-static void ucp_tag_eager_contig_zcopy_completion(uct_completion_t *self)
+static void ucp_tag_eager_contig_zcopy_completion(uct_completion_t *self,
+                                                  ucs_status_t status)
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: why status is not used?

File: src/ucs/type/status.h [hunk 10]
```diff
@@ -42,9 +49,23 @@ typedef enum {
...
+    UCS_ERR_FIRST_LINK_FAILURE     = -40,
+    UCS_ERR_LAST_LINK_FAILURE      = -59,
+    UCS_ERR_FIRST_ENDPOINT_FAILURE = -60,
```
Comment (yosefe) [inline src/ucs/type/status.h]: cannot use "endpoint" terminology here because it's UCS

File: src/uct/api/tl.h [hunk 11]
```diff
@@ -71,6 +67,14 @@ typedef struct uct_iface_ops {
     int          (*iface_is_reachable)(uct_iface_h iface,
                                        const uct_device_addr_t *addr);
 
+    void         (*ep_destroy)(uct_ep_h ep);
```
Comment (yosefe) [inline src/uct/api/tl.h]: why needed to move the functions?

File: src/uct/base/uct_iface.c [hunk 12]
```diff
@@ -194,6 +194,40 @@ static ucs_status_t uct_base_iface_flush(uct_iface_h tl_iface)
...
+
+void uct_base_iface_set_error(uct_iface_h tl_iface, ucs_status_t status)
+{
+    /*
+    void **func_iter;
+    tl_iface->status = status;
+    */
+
+    /* Overwrite functions with interface as first argument */
+    /*for (func_iter = (void**)&tl_iface->ops;
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: ??

File: src/uct/base/uct_iface.c [hunk 13]
```diff
@@ -194,6 +194,40 @@ static ucs_status_t uct_base_iface_flush(uct_iface_h tl_iface)
     return UCS_OK;
 }
 
+/*
+static ucs_status_t uct_iface_nop_iface_arg(uct_iface_h tl_iface)
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: ??

File: src/uct/base/uct_iface.c [hunk 14]
```diff
@@ -216,6 +250,7 @@ UCS_CLASS_INIT_FUNC(uct_iface_t, uct_iface_ops_t *ops)
 
 UCS_CLASS_CLEANUP_FUNC(uct_iface_t)
 {
+    self->status = UCS_OK;
```
Comment (yosefe) [inline src/uct/base/uct_iface.c]: why needed?

File: src/uct/base/uct_iface.h [hunk 15]
```diff
@@ -449,12 +449,26 @@ uct_iface_invoke_am(uct_base_iface_t *iface, uint8_t id, void *data,
...
+ * Set error on interface.
+ *
+ * In the event of an error rendering the interface temporarily or permanently
+ * disfunctional, this function causes further calls to the interface to return
+ * the given error code - except for a call to iface_destroy().
+ *
+ * @param iface  Subject interface.
+ * @param status Error code to set.
+ */
+void uct_base_iface_set_error(uct_iface_h iface, ucs_status_t status);
```
Comment (yosefe) [inline src/uct/base/uct_iface.h]: where should this function be used?
+ please use doxygen style doc

File: src/ucp/tag/eager_snd.c [hunk 16]
```diff
@@ -270,13 +270,15 @@ static ucs_status_t ucp_tag_eager_contig_zcopy_multi(uct_pending_req_t *self)
...
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct_comp);
     ucp_tag_eager_contig_zcopy_req_complete(req);
 }
 
-static void ucp_tag_eager_generic_complere(uct_pending_req_t *self)
+static void ucp_tag_eager_generic_complete(uct_pending_req_t *self,
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: why is this change needed? the PR should be for UCT

File: src/ucp/tag/eager_snd.c (hunk 8 above)
Comment (alex--m) [inline src/ucp/tag/eager_snd.c]: yes, but this commit is only about the API, and this change was made only for the sake of a successful build. Next phase will be the implementation, where such issues will be addressed.

Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: but this is UCP and the PR is for UCT - so why is this needed?

File: src/uct/api/tl.h (hunk 11 above)
Comment (alex--m) [inline src/uct/api/tl.h]: see comment above. basically so that funcs are grouped by first arg.

File: src/ucp/tag/eager_snd.c (hunk 9 above)
Comment (alex--m) [inline src/ucp/tag/eager_snd.c]: see prev. comment.

File: src/uct/base/uct_iface.c (hunk 12 above)
Comment (alex--m) [inline src/uct/base/uct_iface.c]: This is a reference for future implementation, complementing the comment. I don't mind removing it, but I think it clarifies the functionality I'm suggesting for set_error().

File: src/uct/base/uct_iface.c (hunk 13 above)
Comment (alex--m) [inline src/uct/base/uct_iface.c]: used in set_error()

File: src/uct/base/uct_iface.c (hunk 14 above)
Comment (alex--m) [inline src/uct/base/uct_iface.c]: Leftovers after rebase... sorry about that :)

File: src/uct/base/uct_iface.h (hunk 15 above)
Comment (alex--m) [inline src/uct/base/uct_iface.h]: The function is to be used wherever an "persistent" error occurs - and subsequent calls must return the same error. I didn't put in the calls so that the build may pass. This will be done in the next phase of development.

File: src/uct/api/tl.h (hunk 11 above)
Comment (yosefe) [inline src/uct/api/tl.h]: but `uct_base_iface_set_error()` is currently not doing anything - so we should postpone the change (and the comment) to the implementation PR

File: src/uct/base/uct_iface.c (hunk 12 above)
Comment (yosefe) [inline src/uct/base/uct_iface.c]: so it should go to next PR

File: src/uct/base/uct_iface.c (hunk 13 above)
Comment (yosefe) [inline src/uct/base/uct_iface.c]: pls remove all commented-out code

File: src/uct/base/uct_iface.h (hunk 15 above)
Comment (yosefe) [inline src/uct/base/uct_iface.h]: this is not part of UCT API so it should go to next PR

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1761/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1154/ for details (Mellanox internal link).

**Comment by alex--m** [comment]
@yosefe How would you propose to approach the following?
I changed uct_def.h to add status for callbacks:
```
typedef void (*uct_completion_callback_t)(uct_completion_t *self,
                                          ucs_status_t status);
```
If the patch is limited to UCT - I can't change use of ucp_proto_t in UCP, where an incompatible function pointer is passed, and the build fails (for example ucp_tag_eager_contig_zcopy_completion).

**Comment by yosefe** [comment]
@alex--m need to change only the UCP functions which are passed directly to UCT as the callback of this type.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1762/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1155/ for details (Mellanox internal link).

File: src/ucp/tag/eager_snd.c (hunk 9 above)
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: ok

File: src/uct/api/tl.h [hunk 17]
```diff
@@ -159,6 +159,7 @@ typedef struct uct_iface_ops {
  */
 typedef struct uct_iface {
     uct_iface_ops_t          ops;
+    ucs_status_t             status;
```
Comment (yosefe) [inline src/uct/api/tl.h]: looks like this field is unused now

**Comment by yosefe** [comment]
@alex--m please squash the commits after applying CR comments

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1192/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1799/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1199/ for details (Mellanox internal link).

File: src/ucs/type/status.h [hunk 18]
```diff
@@ -42,9 +49,23 @@ typedef enum {
...
+    UCS_ERR_LAST_ENDPOINT_FAILURE  = -79,
+    UCS_ERR_ENDPOINT_TIMEOUT       = -80,
+
     UCS_ERR_LAST                   = -100
 } UCS_S_PACKED ucs_status_t ;
 
+#define USC_IS_LINK_ERROR(code) \
```
Comment (yosefe) [inline src/ucs/type/status.h]: 1. please use leading _ for variable names
2. please surround _code with () when using inside the macro

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1837/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1228/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Comment by yosefe** [comment]
@shamisp  pls review

**Comment by shamisp** [comment]
👍
